change to .phpcs.xml.dist and update rules

NicktheGeek opened the pull request:
The .phpcs.xml.dist rule file is detected and implemented automatically, reducing complexity for running phpcs.

The updated rules account for new wpcs sniff properties including the ability to check global prefix and i18n implementation with the correct text domain when specified.

The values have been left blank so the sniffs are generic, but the rules are in place to make it easier to get started by scaffold.

This is tangentially related to https://github.com/wp-cli/scaffold-command/issues/109

This is also loosely related to https://github.com/wp-cli/scaffold-command/issues/63 as updating composer scaffolds could work with this change.

@GaryJones may be able to provide additional feedback on this file change.

schlessera commented:
Yes, I would indeed love feedback from @GaryJones and @jrfnl to know whether this would be a good default starting point for new plugins.

schlessera commented:
@NicktheGeek For the blanks, it would be helpful to add comments so that developers immediately know where to turn the dials.

NicktheGeek commented:
@schlessera good idea.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
+    <arg name="extensions" value="php"/>
+    <arg name="report" value="full"/>
```
Remove the reports args - the default is a full report anyway, and any different reports should be decided upon at runtime.

jrfnl replied:
FYI: I totally agree with @GaryJones' advice about the report.

GaryJones replied:
Heh - the advice is what I took from Juliette!

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0"?>
+<ruleset name="WordPress Coding Standards for Plugins">
```
This naming is a bit misleading - perhaps `"WordPress Coding Standards for this plugin"? is better, and less implies that there's a WPCS specifically for Plugins?

NicktheGeek replied:
I'm pulling this name from the current version, but I am good to do this naming change.

jrfnl replied:
May I suggest "_WPCS based custom ruleset for plugins_" or "_WordPress Coding Standards based custom ruleset for your plugin/for plugin xyz_" ?

Makes it very clear that this is a _custom ruleset_ and not a _standard_. There is a significant difference between those two and naming this right will help people find the right information if they would search for more info.

To explain the difference:
* A **standard** can have add new sniffs, can be installed and referenced by name.
* A **custom ruleset** uses existing standards and ruleset and may add configuration for these.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
+    <arg name="extensions" value="php"/>
```
Another potentially useful arg is `<arg name="parallel" value="8"/>` - the number can vary, but 8 is a nice starting point.

jrfnl replied:
:+1: Comment to add along-side it: `<!-- Enabled parallel processing when available for faster results. -->`

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors" />
```
Remove the space before the slash.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="."/><!-- Strip the file paths down to the relevant bit -->
```
I know @jrfnl favours `./` here rather than `.`.

jrfnl replied:
Either is fine as long as the `[.]phpcs.xml.dist` file is located in the project root.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+
+	<!-- Rules: Check PHP version compatibility -->
+	<config name="testVersion" value="5.3-"/>
+	<rule ref="PHPCompatibility"/>
```
Since this is a WordPress plugin/theme, then it can reference the `PHPCompatibilityWP` ruleset instead - would need the `composer.json` template updated as well to pull that package in.

NicktheGeek replied:
Since that is outstanding, should this remain as is and implement a change for it with #63?

jrfnl replied:
I would most definitely recommend changing this already to use the `PHPCompatibilityWP` standard instead as it will prevent false positives for things WP back-fills/poly-fills.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+	<rule ref="PHPCompatibility"/>
+
+    <!-- Rules: WordPress Coding Standards -->
+    <config name="minimum_supported_wp_version" value="3.3"/>
```
This seems particularly low. I personally use current - 2 versions (so, 4.7 right now), and WPCS default is current - 3 versions (so, 4.6 for WPCS 1.0.0).

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <!-- Rules: WordPress Coding Standards -->
+    <config name="minimum_supported_wp_version" value="3.3"/>
+    <rule ref="WordPress">
+        <exclude name="WordPress.VIP"/>
```
Might be worth waiting until WPCS 1.0.0 drops, which already removes `WordPress-VIP` from the `WordPress` ruleset.

jrfnl replied:
This will work either way, even if the WPCS VIP deprecation would not make it into 1.0.0, so I'd leave it for now.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <rule ref="WordPress.NamingConventions.PrefixAllGlobals">
+        <properties>
+            <!-- Value: provide the function, class, and variable prefixes used. Separate multiple prefixes with a comma. -->
+            <property name="prefixes" type="array" value=""/>
```
@jrfnl may know better, but having a `my-plugin` placeholder here may better point folks towards changing it to something useful when it gives some warnings on the first run, rather than an empty value which may either blow things up, or not show any related violations (and therefore not drawing attention to changing it) at all.

jrfnl replied:
> having a my-plugin placeholder here may better point folks towards changing it to something useful when it gives some warnings on the first run

I concur with @GaryJones, please do add a placeholder to give people incentive to change it.

> may either blow things up, or not show any related violations

An empty value won't blow things up, but will basically disable the sniff.

GaryJones commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,44 @@
+    <rule ref="WordPress.WP.I18n">
+        <properties>
+            <!-- Value: provide the text domain used. -->
+            <property name="text_domain" type="array" value=""/>
```
I think WPCS 1.0.0 now supports a better way of adding in array properties - individual entries, instead of a comma-separated value.

jrfnl replied:
Not WPCS, but PHPCS 3.3.0+. So if you would want to use the new array format, you would need to make sure that people are aware that they will need a very recent PHPCS version.

NicktheGeek replied:
@jrfnl will the current CSV method work with PHPCS 3.3.0+? If so it seams like it might be better to wait before implementing that even by reference.

GaryJones commented:
Rushing the review, and @jrfnl can fill in some blanks (or I can another day), but overall +1 for the PR to make it a much nicer starting point to implementing WPCS.

Thanks @NicktheGeek!

NicktheGeek commented:
thanks you @GaryJones after you helped me get this on GFWA I've used it on several projects and I can't tell you the number of times it has found a minor typo in the text domain and saved me so much headache down the road. That one feature is worth the price of admission.

NicktheGeek commented:
@jrfnl Thanks! I've added the changes you recommended.

@schlessera this may need some additional feedback from @jrfnl or @GaryJones but so far I've implemented almost everything suggested. There are a couple of comments regarding WPCS 1.0.0 but IMO we are better if this is implemented now and an issue created to update when 1.0.0 comes out with additional improvements. I'm good either way though pending release timeline for WP CLI 2.0 and WPCS 1.0.0.

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+
+    <!-- What to scan -->
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
```
Be aware that `exclude-pattern` values are interpreted as regular expressions, so this snippet would exclude *every* directory which included the string `vendor/`, i.e. `/some-other-vendor/` would also be excluded.

While this may be something which will rarely cause issues, you can stabilize this to just the Composer `vendor` directory by using something along the lines of `^/vendor/*` or the slightly more generic `./vendor/`

jrfnl replied:
For more information about how the `exclude-pattern`s are interpreted, see: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-files-and-folders

jrfnl replied:
Darn... my bad. The more generic version should be `/vendor/`. The `.` which I'd inadvertently placed in front of it will not be interpreted as root, but as a regex wildcard, so, while it shouldn't give any problems, it also isn't necessary.

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+    <file>.</file>
+    <exclude-pattern>vendor/</exclude-pattern>
+
+    <!-- How to scan -->
```
I would recommend adding a link here to the PHPCS wiki so people will know where to find more information about these options.

You can either use the generic wiki link: https://github.com/squizlabs/PHP_CodeSniffer/wiki

Or links to both these pages as good starting points for people:
* Usage instructions: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Usage
* Annotated ruleset: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+    <arg name="extensions" value="php"/>
+    <arg name="parallel" value="8"/><!-- Enables parallel processing when available for faster results. -->
+
+	<!-- Rules: Check PHP version compatibility -->
```
Indentation is off here. Tabs vs spaces.

jrfnl replied:
As per the earlier comments, using `PHPCompatibilityWP` will be the better choice.

I'd also recommend adding a link to the PHPCompatibilityWP repo as a comment, again to help people to more easily find out more info: https://github.com/PHPCompatibility/PHPCompatibilityWP

Maybe also add a link for more info about what the `testVersion` means and how it works: https://github.com/PHPCompatibility/PHPCompatibility#sniffing-your-code-for-compatibility-with-specific-php-versions

You may also want to add a comment saying that people can change the `testVersion` to suit their project.

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+	<config name="testVersion" value="5.3-"/>
+	<rule ref="PHPCompatibility"/>
+
+    <!-- Rules: WordPress Coding Standards -->
```
Along the same lines as some of my other comments, you may want to add a link here to the WPCS repo and/or to the WPCS wiki page about configuring sniffs:
* https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards
* https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+            <property name="text_domain" type="array" value="my-plugin"/>
+        </properties>
+    </rule>
+    <rule ref="WordPress.WhiteSpace.ControlStructureSpacing">
```
This is more a personal preference, than a recommended setting. I'm not sure if it belong in a scaffold.

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+        <properties>
+            <property name="blank_line_check" value="true"/>
+        </properties>
+    </rule>
```
I haven't looked at the scaffold in detail, but in case it also adds a PHPUnit test set-up, you may want to add an additional set of custom properties for that: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#custom-unit-test-classes

jrfnl commented on templates/.phpcs.xml.dist:
```diff
@@ -0,0 +1,42 @@
+    <!-- How to scan -->
+    <arg value="sp"/> <!-- Show sniff and progress -->
+    <arg name="basepath" value="./"/><!-- Strip the file paths down to the relevant bit -->
+    <arg name="colors"/>
```
Just a side-note: `colors` will only work on certain platforms. Not sure by heart, but IIRC it works on Linux and Mac, but not on Windows.

jrfnl commented:
Forgot to respond to this:

> There are a couple of comments regarding WPCS 1.0.0 but IMO we are better if this is implemented now and an issue created to update when 1.0.0 comes out with additional improvements.

WPCS 1.0.0 was due to come out last week, but there's still one issue which needs to be fixed. Expect it very soon though.

All the same, for the ruleset as it is, AFAICS, no changes would be needed for WPCS 1.0.0, so you're good.

GaryJones commented:
The failure looks like some behat tests need updating to reference `.phpcs.xml.dist` (leading dot) instead of `phpcs.xml.dist`.

NicktheGeek commented:
@GaryJones @jrfnl will it be ok for this to not have the leading .? I think it is required to be automatically used but I want to make sure I'm understanding the reason.

jrfnl commented:
> @GaryJones @jrfnl will it be ok for this to not have the leading .? I think it is required to be automatically used but I want to make sure I'm understanding the reason.

Do you mean for the actual file name `phpcs.xml.dist` ? (in contrast to my remark about the exclude pattern https://github.com/wp-cli/scaffold-command/pull/161#discussion_r204142662 )

Yes, no problem.

Allowing a leading `.` for the filename was introduced in PHPCS 3.1.0 with a file loading order change in 3.1.1.

So for the custom ruleset with a leading dot to work, the minimum PHPCS requirement would have to be set at 3.1.0.

The file loading order (as of PHPCS 3.1.1) is:
1. `.phpcs.xml`
2. `phpcs.xml`
3. `.phpcs.xml.dist`
4. `phpcs.xml.dist`

In other words, not having the leading `.` now allows three different files to overrule the file.

Refs:
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.0
* https://github.com/squizlabs/PHP_CodeSniffer/releases/tag/3.1.1

GaryJones commented:
> will it be ok for this to not have the leading .?

And beyond what @jrfnl just said, whether it's right for PHPCS 3.1[.1] to be the minimum version for scaffolded plugins is a different matter, though for new plugins I don't see much reason why the default shouldn't be 3.3.

NicktheGeek commented:
I went ahed and updated the tests to solve for the file ext change

jrfnl commented on templates/plugin-distignore.mustache:
```diff
@@ -21,7 +21,7 @@ phpunit.xml.dist
 multisite.xml
 multisite.xml.dist
 phpcs.xml
```
You may also want to add `.phpcs.xml` here and possibly to the `.gitignore` file too ?

NicktheGeek replied:
@jrfnl thanks! I added .phpcs.xml but .gitignore is already in the file higher up.

jrfnl commented:
FYI: in the mean time, both PHPCS 3.3.1 as well as WPCS 1.0.0 have been released.

schlessera commented:
Thanks to @NicktheGeek for the pull request, and thanks to @GaryJones & @jrfnl for the extensive review!
